weed/s3api/auth_signature_v4.go

```diff
@@ -154,236 +157,248 @@ func parseSignV4(v4Auth string) (sv signValues, aec s3err.ErrorCode) {
-		return nil, errCode
+// buildPathWithForwardedPrefix combines forwarded prefix with URL path while preserving trailing slashes.
+// This ensures compatibility with S3 SDK signatures that include trailing slashes for directory operations.
+func buildPathWithForwardedPrefix(forwardedPrefix, urlPath string) string {
+	fullPath := forwardedPrefix + urlPath
+	hasTrailingSlash := strings.HasSuffix(urlPath, "/") && urlPath != "/"
+	cleanedPath := path.Clean(fullPath)
+	if hasTrailingSlash && !strings.HasSuffix(cleanedPath, "/") {
+		cleanedPath += "/"
 	}
+	return cleanedPath
+}
 
-	// Compute payload hash for non-S3 services
-	if signV4Values.Credential.scope.service != "s3" && hashedPayload == emptySHA256 && r.Body != nil {
-		var err error
-		hashedPayload, err = streamHashRequestBody(r, iamRequestBodyLimit)
-		if err != nil {
-			return nil, s3err.ErrInternalError
-		}
-	}
+// v4AuthInfo holds the parsed authentication data from a request,
+// whether it's from the Authorization header or presigned URL query parameters.
+type v4AuthInfo struct {
+	Signature     string
+	AccessKey     string
+	SignedHeaders []string
+	Date          time.Time
+	Region        string
+	Service       string
+	Scope         string
+	HashedPayload string
+	IsPresigned   bool
+}
 
-	// Extract all the signed headers along with its values.
-	extractedSignedHeaders, errCode := extractSignedHeaders(signV4Values.SignedHeaders, r)
+// verifyV4Signature is the single entry point for verifying any AWS Signature V4 request.
+// It handles standard requests, presigned URLs, and the seed signature for streaming uploads.
+func (iam *IdentityAccessManagement) verifyV4Signature(r *http.Request, shouldCheckPermissions bool) (identity *Identity, credential *Credential, calculatedSignature string, authInfo *v4AuthInfo, errCode s3err.ErrorCode) {
+	// 1. Extract authentication information from header or query parameters
+	authInfo, errCode = extractV4AuthInfo(r)
 	if errCode != s3err.ErrNone {
-		return nil, errCode
+		return nil, nil, "", nil, errCode
 	}
 
-	cred := signV4Values.Credential
-	identity, foundCred, found := iam.lookupByAccessKey(cred.accessKey)
+	// 2. Lookup user and credentials
+	identity, cred, found := iam.lookupByAccessKey(authInfo.AccessKey)
 	if !found {
-		return nil, s3err.ErrInvalidAccessKeyID
+		return nil, nil, "", nil, s3err.ErrInvalidAccessKeyID
 	}
 
-	// Extract date, if not present throw error.
-	var dateStr string
-	if dateStr = req.Header.Get("x-amz-date"); dateStr == "" {
-		if dateStr = r.Header.Get("Date"); dateStr == "" {
-			return nil, s3err.ErrMissingDateHeader
+	// 3. Perform permission check
+	if shouldCheckPermissions {
+		bucket, object := s3_constants.GetBucketAndObject(r)
+		action := s3_constants.ACTION_READ
+		if r.Method != http.MethodGet && r.Method != http.MethodHead {
+			action = s3_constants.ACTION_WRITE
+		}
+		if !identity.canDo(Action(action), bucket, object) {
+			return nil, nil, "", nil, s3err.ErrAccessDenied
```

Comment:
I am only conditionally checking permissions now (iam does not, chunck_reader does). This is to conform with the changes in https://github.com/seaweedfs/seaweedfs/pull/7335